h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GamUtils.java

```diff
@@ -0,0 +1,343 @@
+              src_array[colIdx].length);
+    }
+  }
+
+  /***
+   * from Stack overflow by dfa/user1079877
+   * @param fields
+   * @param type
+   * @return
+   */
+  public static List<Field> getAllFields(List<Field> fields, Class<?> type) {
+    fields.addAll(Arrays.asList(type.getDeclaredFields()));
+
+    if (type.getSuperclass() != null) {
+      getAllFields(fields, type.getSuperclass());
+    }
+
+    return fields;
+  }
+
+  public static GLMParameters copyGAMParams2GLMParams(GAMParameters parms, Frame trainData) {
+    GLMParameters glmParam = new GLMParameters();
+    Field[] field1 = GAMParameters.class.getDeclaredFields();
+    setParamField(parms, glmParam, false, field1);
+    Field[] field2 = Model.Parameters.class.getDeclaredFields();
+    setParamField(parms, glmParam, true, field2);
+    glmParam._train = trainData._key;
+    return glmParam;
+  }
+  
+  public static void setParamField(GAMParameters parms, GLMParameters glmParam, boolean superClassParams, Field[] gamFields) {
+    // assign relevant GAMParameter fields to GLMParameter fields
+    List<String> gamOnlyList = Arrays.asList(new String[]{"_k", "_gam_X", "_bs", "_scale", "_train", "_saveZMatrix", 
+            "_saveGamCols", "_savePenaltyMat"});
+    for (Field oneField : gamFields) {
+      try {
+        if (!gamOnlyList.contains(oneField.getName())) {
+          Field glmField = superClassParams?glmParam.getClass().getSuperclass().getDeclaredField(oneField.getName())
+                  :glmParam.getClass().getDeclaredField(oneField.getName());
+          glmField.set(glmParam, oneField.get(parms));
+        }
+      } catch (IllegalAccessException e) { // suppress error printing
+        ;
+      } catch (NoSuchFieldException e) {
+        ;
+      }
+    }
+  }
+
+  public static int locateBin(double xval, double[] knots) {
+    if (xval <= knots[0])  //small short cut
+      return 0;
+    int highIndex = knots.length-1;
+    if (xval >= knots[highIndex]) // small short cut
+      return (highIndex-1);
+
+    int tryBin = -1;
+    int count = 0;
+    int numBins = knots.length;
+    int lowIndex = 0;
+
+    while (count < numBins) {
+      tryBin = (int) Math.floor((highIndex+lowIndex)*0.5);
+      if ((xval >= knots[tryBin]) && (xval < knots[tryBin+1]))
+        return tryBin;
+      else if (xval > knots[tryBin])
+        lowIndex = tryBin;
+      else if (xval < knots[tryBin])
+        highIndex = tryBin;
+
+      count++;
+    }
+    return tryBin;
+  }
+  
+  public static int colIndexFromColNames(String[] colnames, String oneName) {
+    int len = colnames.length;
+    for (int index = 0; index < len; index++)
+      if (colnames[index].equals(oneName))
+        return index;
+      
+      return -1;
+  }
+
+  public static void copyGLMCoeffs2GAMCoeffs(GAMModel model, GLMModel glm, DataInfo dinfo, GLMParameters.Family family,
+                                             int gamNumStart, boolean standardized, int nclass) {
```

Comment:
This method performs the transform of betaZ back to beta as the last bullet of 7.5